src/pymor/analyticalproblems/expressions.py

```diff
@@ -0,0 +1,307 @@
+
+
+class UnaryFunctionCall(Expression):
+
+    numpy_symbol = None
+
+    def __init__(self, arg):
+        self.arg = arg
+        self.shape = arg.shape
+
+    def numpy_expr(self):
+        return f'{self.numpy_symbol}({self.arg.numpy_expr()})'
+
+    def __str__(self):
+        return f'{self.numpy_symbol}({self.arg})'
+
+
+class UnaryReductionCall(Expression):
+
+    def __init__(self, arg):
+        self.arg = arg
+        self.shape = ()
+
+    def numpy_expr(self):
+        return f'{self.numpy_symbol}({self.arg.numpy_expr()}.reshape(input_shape + (-1,)), axis=-1)'
+
+    def __str__(self):
+        return f'{self.numpy_symbol}({self.arg})'
+
+
+def _convert_to_expression(obj):
+    if isinstance(obj, Expression):
+        return obj
+    elif isinstance(obj, Number):
+        return Constant(obj)
+    else:
+        obj = np.array(obj)
+        if obj.dtype == object:
+            raise NotImplementedError
+        return Constant(obj)
+
+
+def _broadcastable(first, second):
+    return all(f == s or f == 1 or s == 1 for f, s in zip(first[::-1], second[::-1]))
+
+
+class Sum(BinaryOp):  numpy_symbol = '+'   # NOQA
+class Diff(BinaryOp): numpy_symbol = '-'   # NOQA
+class Prod(BinaryOp): numpy_symbol = '*'   # NOQA
+class Div(BinaryOp):  numpy_symbol = '/'   # NOQA
+class Pow(BinaryOp):  numpy_symbol = '**'  # NOQA
+class Mod(BinaryOp):  numpy_symbol = '%'   # NOQA
+class LE(BinaryOp):   numpy_symbol = '<='  # NOQA
+class GE(BinaryOp):   numpy_symbol = '>='  # NOQA
+class LT(BinaryOp):   numpy_symbol = '<'   # NOQA
+class GT(BinaryOp):   numpy_symbol = '>'   # NOQA
+
+
+class sin(UnaryFunctionCall):      numpy_symbol = 'sin'      # NOQA
+class cos(UnaryFunctionCall):      numpy_symbol = 'cos'      # NOQA
+class tan(UnaryFunctionCall):      numpy_symbol = 'tan'      # NOQA
+class arcsin(UnaryFunctionCall):   numpy_symbol = 'arcsin'   # NOQA
+class arccos(UnaryFunctionCall):   numpy_symbol = 'arccos'   # NOQA
+class arctan(UnaryFunctionCall):   numpy_symbol = 'arctan'   # NOQA
+class sinh(UnaryFunctionCall):     numpy_symbol = 'sinh'     # NOQA
+class cosh(UnaryFunctionCall):     numpy_symbol = 'cosh'     # NOQA
+class tanh(UnaryFunctionCall):     numpy_symbol = 'tanh'     # NOQA
+class arcsinh(UnaryFunctionCall):  numpy_symbol = 'arcsinh'  # NOQA
+class arccosh(UnaryFunctionCall):  numpy_symbol = 'arccosh'  # NOQA
+class arctanh(UnaryFunctionCall):  numpy_symbol = 'arctanh'  # NOQA
+class exp(UnaryFunctionCall):      numpy_symbol = 'exp'      # NOQA
+class exp2(UnaryFunctionCall):     numpy_symbol = 'exp2'     # NOQA
+class low(UnaryFunctionCall):      numpy_symbol = 'log'      # NOQA
+class log2(UnaryFunctionCall):     numpy_symbol = 'log2'     # NOQA
+class log10(UnaryFunctionCall):    numpy_symbol = 'log10'    # NOQA
+class sqrt(UnaryFunctionCall):     numpy_symbol = 'sqrt'     # NOQA
+class abs(UnaryFunctionCall):      numpy_symbol = 'abs'      # NOQA
+class sign(UnaryFunctionCall):     numpy_symbol = 'sign'     # NOQA
+
```

Comment:
Maybe making all these class defs dynamic to avoid c&p errors like in `low/log` could be a good idea? Would avoid the noqas too.

Reply:
I don't think so. Dynamically created classes are not picklable. Also, I would like to have the option to define symbols whose numpy equivalent has a different name. There will also be further attributes here, once other backends are implemented.

Reply:
Alright. Didn't consider you'd have to implement `__reduce__` then.